pkg/ssl.go

```diff
@@ -0,0 +1,173 @@
+	HttpsHostname      string `yaml:"https_hostname"`
+	HttpsPort          string `yaml:"https_port"`
+}
+
+func NewSSL(instance *Instance) *SSL {
+	return &SSL{
+		instance: instance,
+	}
+}
+
+func (s SSL) Setup(keyStorePassword, trustStorePassword, certificateFile, privateKeyFile, httpsHostname, httpsPort string) (bool, error) {
+	if !pathx.Exists(certificateFile) {
+		return false, fmt.Errorf("%s > certificate file does not exist: %s", s.instance.ID(), certificateFile)
+	}
+	if !pathx.Exists(privateKeyFile) {
+		return false, fmt.Errorf("%s > private key file does not exist: %s", s.instance.ID(), privateKeyFile)
+	}
+
+	lock := s.lock(keyStorePassword, trustStorePassword, certificateFile, privateKeyFile, httpsHostname, httpsPort)
+	check, err := lock.State()
+	if err != nil {
+		return false, fmt.Errorf("%s > failed to check SSL setup: %w", s.instance.ID(), err)
+	}
+	if check.UpToDate {
+		log.Debugf("%s > SSL already set up (up-to-date)", s.instance.ID())
+		return false, nil
+	}
+
+	params := map[string]any{
+		"keystorePassword":          keyStorePassword,
+		"keystorePasswordConfirm":   keyStorePassword,
+		"truststorePassword":        trustStorePassword,
+		"truststorePasswordConfirm": trustStorePassword,
+		"httpsHostname":             httpsHostname,
+		"httpsPort":                 httpsPort,
+	}
+
+	response, err := s.instance.http.
+		RequestFormData(params).
+		SetFiles(map[string]string{
+			"certificateFile": certificateFile,
+			"privatekeyFile":  privateKeyFile,
+		}).
+		Post(SSLSetupPath)
+
+	if err != nil {
+		return false, fmt.Errorf("%s > failed to setup SSL: %w", s.instance.ID(), err)
+	} else if response.IsError() {
+		// If the response is an error, we try to read the error message from the response body.
+		// The resty library does not provide a way to read the response body if the response is an error.
+		body, err := io.ReadAll(response.RawBody())
+		if err != nil {
+			return false, fmt.Errorf("%s > failed to setup SSL: %s, %w", s.instance.ID(), response.Status(), err)
+		}
+		errorMessage, err := findErrorMessage(string(body[:]))
+		if err != nil {
+			return false, fmt.Errorf("%s > failed to setup SSL: %s, %w", s.instance.ID(), response.Status(), err)
+		}
+		return false, fmt.Errorf("%s > failed to setup SSL: %s, %s", s.instance.ID(), response.Status(), errorMessage)
+	}
+
+	if err := lock.Lock(); err != nil {
+		return true, fmt.Errorf("%s > failed to lock SSL setup: %w", s.instance.ID(), err)
+	}
+
+	return true, nil
+}
+
+// From HTML response body, e.g.:
+// <!DOCTYPE html>
+// <html lang='en'>
+// <head>
+// <title>Error</title>
+// </head>
+// <body>
+// <h1>Error</h1>
+// <dl>
+// <dt class='foundation-form-response-status-code'>Status</dt>
+// <dd>500</dd>
+// <dt class='foundation-form-response-status-message'>Message</dt>
+// <dd>Invalid password for existing key store</dd>
+// <dt class='foundation-form-response-title'>Title</dt>
+// <dd>Error</dd>
+// </dl>
```

Comment:
Is there no better way for this?

Reply:
sometimes AEM is nasty ;) I don't think so

Reply:
Nope, accept header made no difference. It's always HTML.